keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,118 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The purpose of this enhancement is to allow for the expansion of persistent volume claims created by StatefulSets. This entails propagating increases to storage requests in `StatefulSets.volumeClaimTemplates` to associated persistent volume claims.
+
+## Motivation
+In Kubernetes v1.11 the persistent volume expansion feature was promoted to beta. This allowed users to expand volumes by editing storage requests in persistent volume claim objects.
+
+Kubernetes creates a persistent volume for each `volumeClaimTemplate` in the `volumeClaimTemplates` component of a StatefulSet. However, it is not possible to expand persistent volumes, created by StatefulSets, by editing the source `volumeClaimTemplate` in the StatefulSet object. Therefore, it is necessary for the user to individually modify all pods' persistent volume claims, by increasing their storage requests, to expand the underlying persistent volumes. This would have to be repeated each time the number of replicas in the StatefulSet object is increased, since new persistent volumes would be created with the original storage request specified in `volumeClaimTemplates` component.
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by updating the `PersistentVolumeClaimResize` admission controller to incorporate validating updates to `volumeClaimTemplates` within a StatefulSet. Specifically the admission controller will now also check if every `volumeClaimTemplate` in a StatefulSet, which is being resized, is associated with a StorageClass that has `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
```

Comment:
as an alternative to putting the volumeTemplate size request increases in the controller history (where they block rollback), the reconcile loop that ensures PVCs exist could also compare requested size between the template and the existing PVC, and reconcile the existing PVC to request an increase. Was this discussed instead?

Reply:
In this [context](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L266) the "updated set" is actually just the set patched with the update ControllerRevision. If the volumeClaimTemplates object is not in the ControllerRevision, the update revision name will not differ from the revision name marked on each pod (in the pods' labels) and thus the controller [will not terminate](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L509) the pod for an update. 
We want the pod to be terminated to update its associated PVCs due to the reasons described in the KEP (to support both online and offline expansion).

So in essence the kep is suggesting to add the volumeClaimTemplates to the controller revision to ensure the pod terminates in accordance with the existing control flow for the statefulset updates, after which we resize the PVC while the pod is down and recreate the pod once the pvc resize suceeds.

If we don't want to include the volumeClaimTemplates in the revision we could probably modify the update control flow to still cause termination of the pod(hopefully without any side effects), but it was [suggested](url) that the volumeClaimTemplates be added to the ControllerRevision earlier.